src/transactionsFactories/smartContractTransactionsFactory.ts

```diff
@@ -80,25 +84,53 @@ export class SmartContractTransactionsFactory {
 
     createTransactionForExecute(options: {
         sender: IAddress,
-        contractAddress: IAddress,
+        contract: IAddress,
         functionName: string,
         gasLimit: BigNumber.Value,
-        args?: any[]
-    }
-    ): DraftTransaction {
+        args?: any[],
+        nativeTransferAmount?: BigNumber.Value,
+        tokenTransfers?: TokenTransfer[]
+    }): DraftTransaction {
         const args = options.args || [];
-        let parts: string[] = [options.functionName];
+        const tokenTransfer = options.tokenTransfers || [];
+        const nativeTransferAmount = options.nativeTransferAmount ?? 0;
+        const numberOfTokens = tokenTransfer.length;
 
-        const preparedArgs = this.argsToDataParts(args, this.abiRegistry?.getEndpoint(options.functionName));
-        parts = parts.concat(preparedArgs);
+        if (nativeTransferAmount && numberOfTokens) {
+            throw new ErrBadUsage("Can't send both native token and ESDT/NFT tokens");
+        }
+
+        let receiver = options.contract;
+        const tokenComputer = new TokenComputer();
+        let transferArgs: string[] = [];
+
+        if (numberOfTokens === 1) {
+            const transfer = tokenTransfer[0];
+
+            if (tokenComputer.isFungible(transfer.token)) {
+                transferArgs = this.dataArgsBuilder.buildArgsForESDTTransfer(transfer);
+            }
+            else {
+                transferArgs = this.dataArgsBuilder.buildArgsForSingleESDTNFTTransfer(transfer, receiver);
+                receiver = options.sender;
+            }
+        }
+        else if (numberOfTokens > 1) {
+            transferArgs = this.dataArgsBuilder.buildArgsForMultiESDTNFTTransfer(receiver, tokenTransfer)
+            receiver = options.sender;
+        }
+
+        transferArgs.push(transferArgs.length ? utf8ToHex(options.functionName) : options.functionName);
+        transferArgs = transferArgs.concat(this.argsToDataParts(args, this.abiRegistry?.getEndpoint(options.functionName)));
 
         return new DraftTransactionBuilder({
             config: this.config,
             sender: options.sender,
-            receiver: options.contractAddress,
-            dataParts: parts,
+            receiver: receiver,
+            dataParts: transferArgs,
```

Comment:
Also rename `transferArgs` to `dataParts`? Since they not only contain the transfer args per-se.

Reply:
Renamed.